fix(ai): Call `onFinish` when message ui stream is cancelled in `toUIMessageStream`

remorses opened the pull request:
Fix https://github.com/vercel/ai/issues/7469
Fix #7900

Call `onFinish` when abort controller aborts, via the TransformStream cancel method: https://developer.mozilla.org/en-US/docs/Web/API/TransformStream#browser_compatibility

That method is currently only available in Node and Deno

grant0417 commented:
Anything blocking here? Also hitting this issue and would love if this was merged.

remorses commented:
The problem was not strictly related to the abort controller but interrupting the async generator instead.

While writing this comment I also found another issue which was related to consuming the stream with await for and doing an early break.

I updated and added a new test file with the reproduction which do not pass without the fix and use the full streamText function to make sure they reproduce the real problem.

Basically the smallest repro is this code:

```ts
const result = streamText({
  model,
  prompt: 'Generate text',
});

const uiStream = result.toUIMessageStream({
  onFinish: onFinishCallback,
  generateMessageId: () => 'msg-async-iter',
});

// Consume only first few chunks using async iteration, then break
let chunkCount = 0;

for await (const chunk of uiStream) {
  chunkCount++;

  if (chunkCount >= 5) {
    break; // This was not triggering the stream flush(). Instead it just stops consuming the stream and onfinish is never called
  }
}
// onFinish is never called
```

This code will cause the generator to stop and the stream to stop being consumed, the problem was in `createAsyncIterableStream` which does not pass a `return` method in the async iterator. return is necessary to cleanup resources when a `await for` loop breaks

I updated the code to cancel the stream when the async generator breaks. With the added cancel method in the `TransformStream` `onFinish` will be called.

Adding cancel in `TransofrmStream` now also handles the case when the user cancels the `UIMessageStreamResponse`, which is another problem that some users could face, some frameworks probably cancel responses when the request abort controller aborts to prevent memory leaks. Smallest repro is this code:

```tsx
const res = result.toUIMessageStreamResponse()
res.body?.cancel()
```

lgrammel commented:
```ts
const result = streamText({
  model,
  prompt: 'Generate text',
});

const uiStream = result.toUIMessageStream({
  onFinish: onFinishCallback,
  generateMessageId: () => 'msg-async-iter',
});

// Consume only first few chunks using async iteration, then break
let chunkCount = 0;

for await (const chunk of uiStream) {
  chunkCount++;

  if (chunkCount >= 5) {
    break; // This was not triggering the stream flush(). Instead it just stops consuming the stream and onfinish is never called
  }
}
// onFinish is never called
```

@remorses isn't this expected behavior? you never fully consume the stream, so onFinish is not called, since you may continue consuming the stream at any point. if you want to abort stream consumption, you could use an abort controller and abort streamText when you break the loop?

lgrammel commented:
@remorses if this is a problem with the asynciterablestream, a good way may be to have extended unit tests coverage for that file so we can see each individual edge case and how it behaves

remorses commented:
> you never fully consume the stream, so onFinish is not called, since you may continue consuming the stream at any point

That is currently not possible with the current implementation. A for await creates a new iterator each time, but that iterator calls stream.getReader() which fails if the stream is still locked by the previous loop. You would need to add `releaseLock` to add support for that case

I think cancelling the stream when you exit the for loop is the better interface. Otherwise it is too easy to create memory leaks where you never consume the stream when using await for.

Do you want me to add more tests for that file in this PR?

remorses commented:
I added more tests for `createAsyncIterableStream`, all results seem intuitive to me

Trying to iterate on the async iterable twice will simply yield no chunks the second time. This is not a breaking change because with the previous implementation it would just fail because of the missing reader.releaseLock()

lgrammel commented:
sorry for the delay; will try to find time to take a look today or tomorrow

lgrammel commented:
As a first step, I have extracted the async iterable stream changed into a separate pr: https://github.com/vercel/ai/pull/8220

Once that lands and this PR is updated, I will continue looking into the other parts.

lgrammel commented on packages/ai/src/ui/convert-to-model-messages.ts:
```diff
@@ -149,12 +149,7 @@ export function convertToModelMessages(
-                    message: `incomplete tool input is not supported: ${part.toolCallId}`,
-                  });
-                } else {
+                if (part.state !== 'input-streaming') {
```
how are the changes in this file and chat.ts above related to the issue? can these changes be separated?

lgrammel replied:
i wonder how to best handle/document this. right now from what i understand input-streaming would be dropped automatically, correct?

lgrammel commented:
`streamText` has an `onFinish` and an `onAbort` callback. The `onFinish` callback receives usage information that is not available / incorrect in the case of abort. Therefore only `onAbort` should be called.

Is there a bug where `onAbort` is not called? `onFinish` is not called by design atm.

remorses commented:
Are you referring to the test I added in the last commit right?

I added this test to make sure I was understanding when StreamText `onFinish` runs and this was the wanted behaviour. Notice this test is not about aborts but errors. I tried running this same test in main and it still passes so it is unrelated to my changes. 

Do you want me to remove that test?

lgrammel commented:
@remorses ideally move as much as possible into separate prs. overload is real so i appreciate minimal prs focussed on 1 problem.

lgrammel commented on packages/ai/src/generate-text/stream-text-onfinish-abort.test.ts:
```diff
@@ -0,0 +1,224 @@
+import { MockLanguageModelV2 } from '../test/mock-language-model-v2';
+import { streamText } from './stream-text';
+
+describe('streamText onFinish with abort', () => {
```
can you integrate this test into stream-text test and in particular into the ui message stream section (since it is specifically for toUIMessageStream)

remorses replied:
moved to existing file

lgrammel commented on packages/ai/src/ui-message-stream/handle-ui-message-stream-finish.test.ts:
```diff
@@ -2,7 +2,7 @@ import {
   convertReadableStreamToArray,
 } from '@ai-sdk/provider-utils/test';
-import { describe, expect, it, vi } from 'vitest';
+import { beforeEach, describe, expect, it, vi } from 'vitest';
```
can remove whole import line

remorses replied:
fixed

lgrammel commented on packages/ai/src/ui-message-stream/handle-ui-message-stream-finish.ts:
```diff
@@ -101,20 +121,13 @@ export function handleUIMessageStreamFinish<UI_MESSAGE extends UIMessage>({
       transform(chunk, controller) {
         controller.enqueue(chunk);
       },
+      // @ts-expect-error cancel is still new and missing from types https://developer.mozilla.org/en-US/docs/Web/API/TransformStream#browser_compatibility
```
we support node 20 and this was added in 21

<img width="821" height="102" alt="image" src="https://github.com/user-attachments/assets/238b25e7-b86d-4ac1-b001-b55b379d4a0c" />

will this cause any issues on node 20?

remorses replied:
No issues, it simply uses the previous behaviour where onFinish is not called

gr2m reviewed (approved):
Great PR, thank you Tommy 💐

gr2m commented:
@remorses could you look into the merge conflict? Sorry for that, we can look into it ourselves but it might take a bit longer
